### DISCUSSION TITLE
Clarification of Prisma.validator

### DISCUSSION THREAD

**Discussion by TomiHiltunen** [prisma/prisma#9945] [Q&A]
I'm using Prisma.validator like in the docs [here](https://www.prisma.io/docs/concepts/components/prisma-client/advanced-type-safety/prisma-validator). The example has a snippet like this:

```
const userEmail = Prisma.validator<Prisma.UserSelect>()({
  email: true,
})
```

However, when I use `Prisma.validator` like this, the `userEmail` type would still be `any`. Has something changed with the functionality that isn't reflected on the documentation? Additionally, is this function supposed to just cast types or do something else with the input as well?

My Prisma versions:
```
Environment variables loaded from .env
prisma                  : 3.2.1
@prisma/client          : 3.2.1
Current platform        : darwin
Query Engine (Node-API) : libquery-engine b71d8cb16c4ddc7e3e9821f42fd09b0f82d7934c (at node_modules/@prisma/engines/libquery_engine-darwin.dylib.node)
Migration Engine        : migration-engine-cli b71d8cb16c4ddc7e3e9821f42fd09b0f82d7934c (at node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine    : introspection-core b71d8cb16c4ddc7e3e9821f42fd09b0f82d7934c (at node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary           : prisma-fmt b71d8cb16c4ddc7e3e9821f42fd09b0f82d7934c (at node_modules/@prisma/engines/prisma-fmt-darwin)
Default Engines Hash    : b71d8cb16c4ddc7e3e9821f42fd09b0f82d7934c
Studio                  : 0.435.0
```

**Comment by austincrim** [comment]
Hey @TomiHiltunen! Thanks for the question.

I can accurately follow the example from the docs as expected. 

![image](https://user-images.githubusercontent.com/32459922/138950732-a2a13975-8404-4f43-90ae-3eb7af03ddc1.png)

Could you share your `schema.prisma` and I can try to recreate the issue?

> Additionally, is this function supposed to just cast types or do something else with the input as well? 

`Prisma.validator` is only a utility function for creating type-safe objects based on your generated Prisma types. It does not do any other operations on the input.

**Reply by TomiHiltunen** [reply]
> I can accurately follow the example from the docs as expected.

When you explicitly type out the object and its fields the `userEmail` will show exactly those as the type. I expected that it would actually show the type like `Prisma.UserSelect`.

Here is an example where the data is coming from the payload of a request. The data can be anything really. In this scenario the typecasting using the validator doesn't seem to work as I expected. The type of the payload is `any`. Perhaps I've misunderstood the purpose of the validator function?

![image](https://user-images.githubusercontent.com/2010468/139019030-d8109fea-6741-455b-8378-61460fb72e9b.png)

**Reply by austincrim** [reply]
what type is `ctx.request.body` ?

**Reply by TomiHiltunen** [reply]
`ctx.request.body` has type `any`

**Reply by austincrim** [reply]
The result of `Prisma.validator` is dependent on the input having a concrete type at compile time. `Prisma.validator` will not do any kind of validation at runtime. If you're looking to do runtime validation, we recommend you use a 3rd party library or write your own validation logic.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
